### DISCUSSION TITLE
Plugin encapsulation

### DISCUSSION THREAD

**Discussion by garretteklof** [fastify/fastify#2735] [Q&A] [answered]
Hello - thanks for the amazing library! Like many, I've been heads-down using Express for years, but this has made writing Nodejs apps exciting again :)

Quick question regarding scope - I've noticed that when writing plugins (routes), something like this:

```javascript
const plugin = (fastify, opts) => {
  fastify.post("/", async function (request, reply) {
    const data = await getSomeData()
    return someSpecialCleaner(data)
  })
}

function someSpecialCleaner(data) {
  // do stuff
}
```

Does work in development, but does not work in production. In production, I have to bring the function into the scope of the plugin itself? Why is that? I'm guessing it has something to do with the encapsulation, but I was just curious on the discrepancy between the two environments. Also, I know I could throw that logic into a hook, or decorate it but it's a one-off so I figured just keeping it as a standalone made most sense. Any direction pointing would be greatly appreciated.

Thanks again!

**Comment by kibertoad** [comment]
Did you read https://github.com/fastify/fastify/blob/master/docs/Encapsulation.md already? Not sure if it answers all of your questions, but it might help.

**Reply by garretteklof** [reply]
Hi @kibertoad - great work with `fastify-funky` btw. Thanks for the link. I did read through that (or a version of it on fastify.io) - and just looking through it quickly, I definitely understand the parent / child relationship between the `root` on down. My question is more what happens differently between `NODE_ENV` with regards to this process. My guess is that in this example since I have not encapsulated this `someSpecialCleaner` function inside the scope of the plugin and we don't use `fastify-plugin` on the routes, this is actually bound somewhere not accessible to the route (or nowhere). I'll of course be moving it to inside the routes plugin context, but that just makes me wonder if I should be making _everything_ a hook/decorator/middleware or if it's still somewhat idiomatic to write logic like this (my guess no -- if it's shareable, make it a plugin. If it's one-off, keep it in the plugin context).

**Reply by garretteklof** [reply]
My question may seem pretty trivial, but I didn't immediately put it in the context because I do a lot of work in the front-end / React world, where I rely on hoisting to organize my code in a way where "stuff that matters" is at the top, and stuff that is secondary and not needed within function scope relegated to the bottom of the file.

**Reply by kibertoad** [reply]
Haha, thank you!
I did a quick check across `fastify` and `fastify-plugin` codebases, and I don't see NODE_ENV being referenced anywhere, so it's either happening elsewhere, or there is something else at play here. Hopefully more experienced `fastify` colleagues will have a better idea with regards to what is happening there.

**Reply by garretteklof** [reply]
🙈 yeah now that I'm thinking of it - this more than likely has nothing to do with `fastify`. It probably has something to do with the compilation. I'm very new to the Typescript world but in my brief four-day foray I've learned there's quite a bit of differences between `ts-node` and `tsc` 😱 I'm guessing something's happening when I compile using `tsc` for production.

**Comment by jsumners** [comment, marked as answer]
Fastify's encapsulation does not deal with any environment variables. The code you have provided is registering an asynchronous plugin without a callback. So it's rather surprising that it ever works. I imagine what you really want is:

```js
const plugin = (fastify, opts, done) => {
  fastify.post("/", async function (request, reply) {
    const data = await getSomeData()
    return someSpecialCleaner(data)
  })

  done()
}

function someSpecialCleaner(data) {
  // do stuff
}
```

**Reply by garretteklof** [reply]
Yeah, I'm an idiot and just forgot to add the `async` at the top. You're right, it has nothing to do with `fastify` and probably transpiling Typescript. Hopefully someday there will be some standardization with that process because coming in from the plain Javascript world - I was pretty shocked at how difficult it was.

**Reply by jsumners** [reply]
All bets are off if you're transpiling things. The code above should work without issue as `someSpecialCleaner` is still within lexical scope of `plugin` as long as they are in the same file as it is written above.

**Reply by garretteklof** [reply]
Sweet. I'm glad all that I thought I knew about Javascript is not being taken from me.

**Comment by garretteklof** [comment]
Thanks both @kibertoad and @jsumners . The response time was amazing. The verdict is that this has nothing to do with encapsulation, and everything to do with user error. I'll be careful out there with my novice Typescript transpilation skills. Thanks again for all the hard work - this framework is the future, baby 🎸 ❗

**Comment by kibertoad** [comment]
You are welcome! Glad you are using fastify.
